### DISCUSSION TITLE
SD (mp3) to analogue

### DISCUSSION THREAD

**Discussion by francesco-miccoli** [pschatzmann/arduino-audio-tools#975] [Q&A] [answered]
Hello everyone. I am trying to use my ESP32 to play an mp3 file stored in a SD card module and send an analogue output to an Adafruit MAX9744 amplifier. I see there is no explicit example for this so I tried putting something together but in the process of doing that I encountered an error. I get _conflicting declaration 'audio_tools::EncodedAudioStream out'_ when trying to compile 

how can I solve this? From what I understand both the codec and analogue are outputs (as in the _streams-memory_mp3-analog_ example. Thanks in advance!
<img width="894" alt="image" src="https://github.com/pschatzmann/arduino-audio-tools/assets/103175066/2e32e059-e1c3-4044-b81f-f8f3e8990f35">

**Comment by LinusHeu** [comment]
You can't have two variables (objects) with the same name (`out`).
(Generally, in C++. This has nothing to do with this particular library.)

**Reply by LinusHeu** [reply]
Oh, looks to me like there is a mistake in https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-stream/streams-memory_mp3-analog/streams-memory_mp3-analog.ino

**Reply by LinusHeu** [reply]
```ino
/**
 * @file stream-memory_mp3-out.ino
 * @author Phil Schatzmann
 * @brief decode MP3 stream and output as out signal
 * @version 0.1
 * @date 2021-01-24
 * 
 * @copyright Copyright (c) 2021
 
 */

// install https://github.com/pschatzmann/arduino-libhelix.git

#include "AudioTools.h"
#include "AudioCodecs/CodecMP3Helix.h"
#include "BabyElephantWalk60_mp3.h"


MemoryStream mp3(BabyElephantWalk60_mp3, BabyElephantWalk60_mp3_len);
AnalogAudioStream out;  // Analog output 
EncodedAudioStream decoder(&out, new MP3DecoderHelix()); // decoder -> out
StreamCopy copier(decoder, mp3);    // copy mp3 -> decoder -> out

void setup(){
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);  

  // begin processing
  auto cfg = out.defaultConfig();
  out.begin(cfg);

  out.begin();
}

void loop(){
  if (mp3) {
    copier.copy();
  } else {
    auto info = out.decoder().audioInfo();
    LOGI("The audio rate from the mp3 file is %d", info.sample_rate);
    LOGI("The channels from the mp3 file is %d", info.channels);
    out.end();
    stop();
  }
}
```
Can you try it like this? Notice, how I renamed `EncodedAudioStream out` to `EncodedAudioStream decoder`

**Comment by pschatzmann** [comment, marked as answer]
It seems that this bug has been introduced by this [change](https://github.com/pschatzmann/arduino-audio-tools/commit/2bd8c2abfe7c2fa6286ec52fea1f1200e4469f5d#diff-17bc758bbe01f0d26e5c1ef76f174c69f935feb631b5189c667a6b0d7c53d3d0)

The correction has been committed...

**Comment by francesco-miccoli** [comment]
I updated the code according to the correction. Still when I connect the speaker I cannot hear anything if not a faint "beep" like sound. What am I missing? Thanks in advance
<img width="927" alt="SCR-20230906-jqwi" src="https://github.com/pschatzmann/arduino-audio-tools/assets/103175066/99e5edb4-46e8-4890-95c0-8f27e3ac9ade">

**Comment by pschatzmann** [comment]
You are the only one who can find this out! 

You need to have some structured testing approach to identify and resolve your issue. Start simple and extend the complexity:

https://github.com/pschatzmann/arduino-audio-tools/wiki/It's-not-working

As a first test you could replace the AnalogAudioStream with a CSVStream to check if you get some valid audio data. 
You did not provide any logging data, so I assume that you see the copy entries...

ps. don't forget to clean up the unused typedf
